### PR TITLE
feature/21-Add functionality to add picture, add and retrieve categories to and from dummy dataset

### LINKED ISSUE #21: feature/21-Add functionality to add picture, add and retrieve categories to and from dataset

### PR DESCRIPTION
**Pull request by martin-lenweiter**
add functionality to datasetInterface and Presenter to allow to put pictures to the dataset, and put and get categories. Also add corresponding dummy implementations

### CODE REVIEW

**Comment by bbjubjub2494** [comment]
Thought: if you add a Drawable attribute to CategorizedPicture, this might prevent its instances from [serializing](https://developer.android.com/reference/java/io/Serializable), which Nicolas [needs](https://github.com/HumanLearning2021/HumanLearningApp/blob/f69a55b4906eb9d10054cb8424448debfe8d8208/app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/view/DisplayImageActivity.kt#L15).

**Comment by martin-lenweiter** [comment]
> Thought: if you add a Drawable attribute to CategorizedPicture, this might prevent its instances from [serializing](https://developer.android.com/reference/java/io/Serializable), which Nicolas [needs](https://github.com/HumanLearning2021/HumanLearningApp/blob/f69a55b4906eb9d10054cb8424448debfe8d8208/app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/view/DisplayImageActivity.kt#L15).

Good thought. We can discuss this tomorrow the 3 of us. Wouldn't be difficult to change the code to accommodate this.

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/presenter/DummyUIPresenter.kt
Comment (bbjubjub2494) [inline app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/presenter/DummyUIPresenter.kt]: Can we take the opportunity to solve this warning?
```
The corresponding parameter in the supertype 'UIPresenter' is named 'categoryName'. This may cause problems when calling this function with named arguments.
```
```suggestion
    override suspend fun getPicture(categoryName: String): CategorizedPicture? {
```

**Comment by martin-lenweiter** [comment]
I'm thinking that DummyCategory should actually be encapsulated in the DummyDatasetInterface, as this is closer to how the real model is gonna behave. i.e. no one outside the model should be able to create DummyCategory objects. The only way they have access to a such objects is by calling putCategory(). @lourkeur  @NicolasVial what do you two think? Nicolas I see that in your tests you make use of the DummyCategory constructor.

Edit: same applies to DummyCategorizedPicture

**Comment by bbjubjub2494** [comment]
> I'm thinking that DummyCategory should actually be encapsulated in the DummyDatasetInterface, as this is closer to how the real model is gonna behave. i.e. no one outside the model should be able to create DummyCategory objects. The only way they have access to a such objects is by calling putCategory(). @lourkeur @NicolasVial what do you two think? Nicolas I see that in your tests you make use of the DummyCategory constructor.
> 
> Edit: same applies to DummyCategorizedPicture

I think that's defensible from a design perspective, but if it breaks everyone's tests it's unnecessary. Also do you intend to make it nested? private? internal?

**Comment by martin-lenweiter** [comment]
> > I'm thinking that DummyCategory should actually be encapsulated in the DummyDatasetInterface, as this is closer to how the real model is gonna behave. i.e. no one outside the model should be able to create DummyCategory objects. The only way they have access to a such objects is by calling putCategory(). @lourkeur @NicolasVial what do you two think? Nicolas I see that in your tests you make use of the DummyCategory constructor.
> > Edit: same applies to DummyCategorizedPicture
> 
> I think that's defensible from a design perspective, but if it breaks everyone's tests it's unnecessary. Also do you intend to make it nested? private? internal?

Agreed. I was thinking of making them both inner classes (from what I've understood only the parent class is allowed to create instances of inner classes). For now let's keep it as is, but depending on how much we end up using the dummy data set we should consider changing it.

**Comment by martin-lenweiter** [comment]
Make it a draft, as substantial merging work will have to be done once Nicolas and Louis' PR are merged.

**Review by Nitwix — APPROVED** [review]
Seems all good to me! Should it still be a draft or is it ready to be merged?

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DummyCategorizedPicture.kt
Comment (martin-lenweiter) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DummyCategorizedPicture.kt:15]: It is needed in @lourkeur's implementation, I don't remember exactly why

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/presenter/UIPresenter.kt
Comment (martin-lenweiter) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/presenter/UIPresenter.kt]: I didn't do much research on the presenter's role, but actually this seems true. So indeed there's no need for an interface.

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DataSetInterface.kt
Comment (bbjubjub2494) [inline app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DataSetInterface.kt:41]: A category's key is not its name in my book, so that constraint is weird to implement for me. It's more natural to create a duplicate category. Obviously the dummy implementation will have problems w/ that though.

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DummyDataSetInterface.kt
Comment (bbjubjub2494) [inline app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DummyDataSetInterface.kt]: What do I do w/ a Serializable? this needs to be a more specific type, or a different one altogether.

EDIT: this applies to the interface, not just this implementation

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DataSetInterface.kt
Comment (martin-lenweiter) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DataSetInterface.kt:41]: Not sure I get what you mean. The name here is just the parameter to the function, so components calling the function might want to get a category that corresponds to a particular name. Or do you want them to refer to categories only by their key, which the `putCategory()` function would return?

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreDatasetInterface.kt
Comment (bbjubjub2494) [inline app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreDatasetInterface.kt]:
```suggestion
    override suspend fun putPicture(picture: android.net.Uri, category: Category): CategorizedPicture {
```
That's what AddPictureActivity returns and I can work with that.

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DummyDataSetInterface.kt
Comment (martin-lenweiter) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DummyDataSetInterface.kt]: a Bitmap maybe? or a reference to one?

**Review by bbjubjub2494 — CHANGES_REQUESTED** [review]
just that one thing

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DataSetInterface.kt
Comment (bbjubjub2494) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DataSetInterface.kt:41]: > The name here is just the parameter to the function, so components calling the function might want to get a category that corresponds to a particular name.

There's getCategory for that. They go use that. It does a query. (`SELECT` in SQL parlance). Then they decide if they want to make a new one. `put*` methods are for creating more objects in the database. (`INSERT`)


> Or do you want them to refer to categories only by their key, which the putCategory() function would return?

Ideally yes, otherwise, how do they even know that the given category exists? And what if there's multiple? That's the point of keys.

Comment (martin-lenweiter) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DataSetInterface.kt:41]: > > The name here is just the parameter to the function, so components calling the function might want to get a category that corresponds to a particular name.
> 
> There's getCategory for that. They go use that. It does a query. (`SELECT` in SQL parlance). Then they decide if they want to make a new one. `put*` methods are for creating more objects in the database. (`INSERT`)

Aren't we talking about `getCategory()`?

Comment (martin-lenweiter) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DataSetInterface.kt:41]: which would correspong to a `SELECT * FROM category WHERE category.name = categoryName` query roughly no?

Comment (bbjubjub2494) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DataSetInterface.kt:41]: Yes precisely. Now what I'm complaining about is that currently, putCategory is `SELECT`, then conditionally `INSERT INTO category` based on the wrong assumption that category names should be unique database-wide.

Comment (martin-lenweiter) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DataSetInterface.kt:41]: Ah okay misunderstanding then, I didn't get that you were talking about `putCategory`

Comment (martin-lenweiter) [reply app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DataSetInterface.kt:41]: > Yes precisely. Now what I'm complaining about is that currently, putCategory is `SELECT`, then conditionally `INSERT INTO category` based on the wrong assumption that category names should be unique database-wide.

But then wouldn't it be possible to create a duplicate category with the same parameters (for now name, but we could imagine more), insert it unconditionally into the database with a different key, and return a reference to another category object to the caller?

File: app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DummyDataSetInterface.kt
Comment (bbjubjub2494) [inline app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/model/DummyDataSetInterface.kt:3]: should run code cleanup.

**Review by bbjubjub2494 — COMMENTED** [review]
Since you severed the connection between `UIPresenter` and `DummyUIPresenter` I have to ask: what interface is Niels (and others) programming against? `UIPresenter`? `DummyUIPresenter`?

**Comment by martin-lenweiter** [comment]
> Since you severed the connection between `UIPresenter` and `DummyUIPresenter` I have to ask: what interface is Niels (and others) programming against? `UIPresenter`? `DummyUIPresenter`?

I actually deleted the interface. Since there will be only one presenter per view, it doesn't make sense to generalize the presenter.

**Comment by bbjubjub2494** [comment]
> > Since you severed the connection between `UIPresenter` and `DummyUIPresenter` I have to ask: what interface is Niels (and others) programming against? `UIPresenter`? `DummyUIPresenter`?
> 
> I actually deleted the interface. Since there will be only one presenter per view, it doesn't make sense to generalize the presenter.

ok gotcha.

**Review by bbjubjub2494 — APPROVED** [review]
please squash-merge so that future maintainers don't have to look at this. Otherwise LGTM

**Pull request merged**
